tests/server.rs

```diff
@@ -204,6 +204,32 @@ fn sends_reset_cancel_when_req_body_is_dropped() {
     srv.join(client).wait().expect("wait");
 }
 
+#[test]
+fn sends_goaway_when_serv_closes_connection() {
```

Comment:
I know you didn't originally intend to add graceful shutdown to the server, but it looks like this adds it! It'd be bonus points if there could be another test showing multiple requests being sent and accepted, and then `close_connection()` called before the responses (and their bodies?) are finished, that the outstanding streams can finish before the connection terminates.

(I suppose we'd also need to do something about any new streams sent after sending the goaway, I don't recall what the spec says to do. `REFUSED_STREAM`?)

Reply:
The spec is slightly confusing on this matter. In one case, it says:

> Receivers of a GOAWAY frame MUST NOT open additional streams on the connection

But it doesn't mention a protocol error. It seems to suggest those new streams can simply be discarded:

> After sending a GOAWAY frame, the sender can discard frames for streams initiated by the receiver with identifiers higher than the identified last stream

With a caveat that data frames still need to count towards congestion flow control.

Reply:
I added a test for showing a request being accepted, then the server closing the connection on the next request (and subsequently ignoring the third request).

Not sure how to make sure any properties like congestion flow control still hold true, please let me know if any changes need to be made!